### PR TITLE
Add SQL Server support [CWC-2454]

### PR DESCRIPTION
**Pull request by tflpd**
## Description of the change

This PR enables SQL Server connections to Windows agents utilizing the DB plugin which will soon be transformed to TCP plugin.

## Testing

Add/edit a policy to have the SQL Server verb in it and try to zli connect to one of your Windows agents. Use the localhost:port returned by the zli in your DBeaver client. As a username use `sa` and as a password the password returned when you deploy a windows agent.

**backend branch:** feat/sqlserver
**bzero branch:** 
**charts branch:** 

#### Ready to run system tests?

- [x] Yes

## Relevant release note information

Release Notes: Add SQL Server support

## Related JIRA tickets

Relates to JIRA: CWC-2454

## Relevant HotFix if applicable

Relevant HotFix PR Number:

## Have you considered the security impacts?

Does this PR have any security impact?

- [ ] Yes
- [x] No

If yes, please explain:

Does this PR add a new/modifies the name/arguments of a `zli` command?

- [x] Yes
- [ ] No

If yes, please provide the docs/zli-reference-manual branch reflecting the changes:

### CODE REVIEW

File: src/handlers/connect/connect.command-builder.ts [hunk 1]
```diff
@@ -38,11 +39,19 @@ export function connectCmdBuilder(yargs: yargs.Argv<{}>, targetTypeChoices: stri
             alias: 'n',
             description: 'Specifies the default namespace to use for Kubernetes connections'
         })
+        .option('action', {
+            type: 'string',
+            choices: verbTypeChoices,
+            demandOption: false,
+            alias: 'a',
+            description: 'Specifies the action to be taken to the specified target'
```
Comment (johncmerfeld) [inline src/handlers/connect/connect.command-builder.ts:47]: "to" -> "on" perhaps?

Comment (tflpd) [reply src/handlers/connect/connect.command-builder.ts]: 08a08b09..1020e2f2

File: src/handlers/connect/connect.command-builder.ts [hunk 2]
```diff
@@ -38,11 +39,19 @@ export function connectCmdBuilder(yargs: yargs.Argv<{}>, targetTypeChoices: stri
...
         .example('$0 connect target-user@target-name.environment-name', 'Connect to targets with the same name in different environment by environment name')
         .example('$0 connect target-user@target-name.environment-GUID', 'Connect to targets with the same name in different environment by environment UUID')
         .example('$0 connect ssm-user@neat-target', 'SSM connect example, uniquely named ssm target')
         .example('$0 connect --targetType dynamic ssm-user@my-dat-config', 'DAT connect example with a DAT configuration whose name is my-dat-config')
         .example('$0 connect admin@neat-cluster --targetGroup system:masters', 'Connect to neat-cluster as the admin Kube RBAC user in the system:masters group')
-        .example('$0 connect admin@neat-cluster -n qa', 'Connect to neat-cluster as the admin Kube RBAC user and set default namespace to qa');
+        .example('$0 connect admin@neat-cluster -n qa', 'Connect to neat-cluster as the admin Kube RBAC user and set default namespace to qa')
+        .example('$0 connect my-rdp-target -action rdp', 'Connect to my-rdp-target using RDP');;
```
Comment (johncmerfeld) [inline src/handlers/connect/connect.command-builder.ts:56]: Maybe `my-windows-target`, to demonstrate that it can be used to disambiguate?

Comment (tflpd) [reply src/handlers/connect/connect.command-builder.ts]: 1020e2f2..fc2a217f

File: src/handlers/connect/tcp-app-connect.handler.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+    const bzeroTargetService = new BzeroTargetHttpService(configService, logger);
+    const bzTarget = await bzeroTargetService.GetBzeroTarget(createUniversalConnectionResponse.targetId);
+
+    // Set our local host
+    const localHost = getOrDefaultLocalhost(null);
+
+    // Make sure we have set our local daemon port
+    let localPort = await getOrDefaultLocalport(defaultPort);
+    // If the chosen port is not the same as the one stored, store the new port
```
Comment (johncmerfeld) [inline src/handlers/connect/tcp-app-connect.handler.ts:68]: I've always found this function a little counterintuitive but I get that this is how it's intended to be used. What if you changed the comment to better reflect the reality? I would rather it say something like:
```
// if a local port hasn't been assigned to this target yet, assign a new one
```

File: src/handlers/connect/tcp-app-connect.handler.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                    type: 'sqlserver',
+                    name: bzTarget.name,
+                    localHost: localHost,
+                    localPort: localPort,
+                    localPid: daemonProcess.pid,
+                    controlPort: runtimeConfig['CONTROL_PORT'],
+                };
+                break;
+            }
+
```
Comment (johncmerfeld) [inline src/handlers/connect/tcp-app-connect.handler.ts:145]: I think you could cut some code if you combined this with the switch statement below (at line 155)

**Review by johncmerfeld — COMMENTED** [review]
Looking good! I'll test later today and approve

File: src/handlers/connect/tcp-app-connect.handler.ts [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    const bzeroTargetService = new BzeroTargetHttpService(configService, logger);
+    const bzTarget = await bzeroTargetService.GetBzeroTarget(createUniversalConnectionResponse.targetId);
+
+    // Set our local host
+    const localHost = getOrDefaultLocalhost(null);
+
+    // Make sure we have set our local daemon port
+    let localPort = await getOrDefaultLocalport(defaultPort);
```
Comment (johncmerfeld) [inline src/handlers/connect/tcp-app-connect.handler.ts:67]: Oh wait a minute, does this adequately handle the case where the random port is taken? This function won't check if the port we pass in is actually available

**Review by johncmerfeld — APPROVED** [review]
Still have some questions about the approach we take with port conflicts, but otherwise yeah this is great. I tested:
  - Connecting with/without various forms of policy access
  - Verifying that the port persists between connections
  - Verifying that `customPort` still works
All working great!

File: src/handlers/connect/tcp-app-connect.handler.ts (hunk 3 above)
Comment (tflpd) [reply src/handlers/connect/tcp-app-connect.handler.ts:68]: Hm, I do not necessarily agree. That comment is a super set of the suggested one. The code will also change the stored one to a new one if the stored one is unavailable

File: src/handlers/connect/tcp-app-connect.handler.ts (hunk 4 above)
Comment (tflpd) [reply src/handlers/connect/tcp-app-connect.handler.ts]: true true
fc2a217f..e09710ae

File: src/handlers/connect/connect.handler.ts [hunk 6]
```diff
@@ -50,21 +50,21 @@ export async function connectHandler(
             envName: parsedTarget.envName,
             targetUser: targetUser,
             targetGroups: argv.targetGroup,
-            targetType: parseTargetType(argv.targetType)
+            targetType: parseTargetType(argv.targetType),
+            verbType: parseVerbType(argv.action)
```
[the remark this answers is not in the document]
Comment (tflpd) [reply src/handlers/connect/connect.handler.ts:54]: Yup, or an exception if multiple/none are available

File: src/handlers/connect/tcp-app-connect.handler.ts (hunk 5 above)
Comment (tflpd) [reply src/handlers/connect/tcp-app-connect.handler.ts:67]: We do that check right below at `checkIfPortAvailable`.

Okay, so what is the expected behavior?

File: src/handlers/connect/tcp-app-connect.handler.ts (hunk 3 above)
Comment (johncmerfeld) [reply src/handlers/connect/tcp-app-connect.handler.ts:68]: Ah, good point, but then I don't think that's the desired behavior, per Mike's slack message

File: src/handlers/connect/tcp-app-connect.handler.ts (hunk 5 above)
Comment (johncmerfeld) [reply src/handlers/connect/tcp-app-connect.handler.ts:67]: My understanding of Mike's ask is that in this case:
  - Fail to connect
  - Tell them they're *probably* already connected
  - Tell them they can use `--customPort`

File: src/handlers/connect/tcp-app-connect.handler.ts (hunk 3 above)
Comment (tflpd) [reply src/handlers/connect/tcp-app-connect.handler.ts]: 23b8ea44..c838b544

File: src/handlers/connect/tcp-app-connect.handler.ts (hunk 5 above)
Comment (tflpd) [reply src/handlers/connect/tcp-app-connect.handler.ts:67]: Yup, that is what the code is doing right now

**Review by johncmerfeld — APPROVED** [review]

File: src/handlers/connect/connect.command-builder.ts [hunk 7]
```diff
@@ -38,11 +39,19 @@ export function connectCmdBuilder(yargs: yargs.Argv<{}>, targetTypeChoices: stri
...
         .example('$0 connect target-user@target-name.environment-name', 'Connect to targets with the same name in different environment by environment name')
         .example('$0 connect target-user@target-name.environment-GUID', 'Connect to targets with the same name in different environment by environment UUID')
         .example('$0 connect ssm-user@neat-target', 'SSM connect example, uniquely named ssm target')
         .example('$0 connect --targetType dynamic ssm-user@my-dat-config', 'DAT connect example with a DAT configuration whose name is my-dat-config')
         .example('$0 connect admin@neat-cluster --targetGroup system:masters', 'Connect to neat-cluster as the admin Kube RBAC user in the system:masters group')
-        .example('$0 connect admin@neat-cluster -n qa', 'Connect to neat-cluster as the admin Kube RBAC user and set default namespace to qa');
+        .example('$0 connect admin@neat-cluster -n qa', 'Connect to neat-cluster as the admin Kube RBAC user and set default namespace to qa')
+        .example('$0 connect my-windows-target -protocol rdp', 'Connect to my-rdp-target using RDP');;
```
Comment (ymarcus93) [inline src/handlers/connect/connect.command-builder.ts:56]: I think you meant `-p` or `--protocol`

Comment (tflpd) [reply src/handlers/connect/connect.command-builder.ts]: c838b544..16f10109

File: src/utils/daemon-utils.ts [hunk 8]
```diff
@@ -478,7 +478,7 @@ export async function shutDownLocalPortAndPid(daemon: DaemonConfig, localPort: n
 export async function checkIfPortAvailable(port: number) {
     const isPortInUse = await checkTcpPort(port, 'localhost');
     if (isPortInUse) {
-        throw new Error(`It looks like an application is using port: ${port}`);
+        throw new Error(`It looks like an application is using port: ${port}. A custom port can be specified using the -customPort flag.`);
```
[the remark this answers is not in the document]
Comment (tflpd) [reply src/utils/daemon-utils.ts]: c838b544..16f10109

**Pull request merged**